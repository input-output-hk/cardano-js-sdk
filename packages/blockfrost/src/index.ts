export * from './blockfrostAssetProvider';
export * from './blockfrostTxSubmitProvider';
export * from './blockfrostNetworkInfoProvider';
export * from './blockfrostUtxoProvider';
export * from './blockfrostChainHistoryProvider';
export * from './blockfrostRewardsProvider';
export { Options } from '@blockfrost/blockfrost-js/lib/types';
export { BlockFrostAPI } from '@blockfrost/blockfrost-js';
