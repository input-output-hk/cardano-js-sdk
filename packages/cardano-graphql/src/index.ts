export { WalletProvider, StakePoolSearchProvider } from '@cardano-sdk/core';
export * from './DgraphClient';
// Do not export this. Using core types elsewhere in the sdk.
// export * as Schema from './Schema';
