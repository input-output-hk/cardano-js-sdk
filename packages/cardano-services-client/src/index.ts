export { TxSubmitProvider } from '@cardano-sdk/core';
export * from './HttpProvider';
export * from './TxSubmitProvider';
