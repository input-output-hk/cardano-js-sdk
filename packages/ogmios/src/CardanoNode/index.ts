export * from './OgmiosCardanoNode';
export * from './OgmiosObservableCardanoNode';
