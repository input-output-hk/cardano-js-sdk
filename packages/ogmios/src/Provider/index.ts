export * from './TxSubmitProvider/index.js';
