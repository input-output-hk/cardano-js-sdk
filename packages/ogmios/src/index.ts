export * from './Provider';
export * from './CardanoNode';
export * from './util';
export * as ogmiosToCore from './ogmiosToCore';
export * as Ogmios from '@cardano-ogmios/client';
