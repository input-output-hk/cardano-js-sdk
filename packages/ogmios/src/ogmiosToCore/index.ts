export * from './block';
export * from './genesis';
export * from './types';
