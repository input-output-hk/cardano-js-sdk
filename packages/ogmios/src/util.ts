import { Cardano } from '@cardano-sdk/core';
import {
  ConnectionConfig,
  InteractionContext,
  InteractionType,
  ServerHealth,
  createInteractionContext
} from '@cardano-ogmios/client';
import { Logger } from 'ts-log';

/**
 * Converts an Ogmios connection URL to an Ogmios ConnectionConfig Object
 *
 * @param {URL} connectionURL Ogmios connection URL
 * @returns {ConnectionConfig} the ConnectionConfig Object
 */
export const urlToConnectionConfig = (connectionURL?: URL): ConnectionConfig => ({
  host: connectionURL?.hostname,
  port: connectionURL ? Number.parseInt(connectionURL.port) : undefined,
  tls: connectionURL?.protocol === 'wss'
});

type CreateInteractionContextOptions = {
  connection?: ConnectionConfig;
  interactionType?: InteractionType;
};

/**
 * Creates an Ogmios InteractionContext with close logic
 *
 * @param {Logger} logger Logger instance
 * @param {CreateInteractionContextOptions} options Options passed to the createInteractionContext function
 * @param {Function} onUnexpectedClose Optional callback to trigger an action when an unexpected close event occurs
 * @returns {Promise<InteractionContext>} Promise resolving an Ogmios InteractionContext
 */
export const createInteractionContextWithLogger = (
  logger: Logger,
  options?: CreateInteractionContextOptions,
  onUnexpectedClose?: () => Promise<void>
): Promise<InteractionContext> =>
  createInteractionContext(
    (error) => {
      logger.error({ error }, error.message);
    },
    async (code, reason) => {
      if (code === 1000) {
        logger.info({ code }, reason);
      } else {
        logger.error({ code }, 'Connection closed');
        await onUnexpectedClose?.();
      }
    },
    options
  );

export const ogmiosServerHealthToHealthCheckResponse = ({ lastKnownTip, networkSynchronization }: ServerHealth) => ({
  localNode: {
    ledgerTip: lastKnownTip,
    networkSync: Cardano.Percent(networkSynchronization)
  },
  ok: networkSynchronization > 0.99
});
