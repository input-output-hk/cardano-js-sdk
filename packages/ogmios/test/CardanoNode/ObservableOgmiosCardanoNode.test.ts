/* eslint-disable sonarjs/no-duplicate-string */
import { CardanoNodeErrors, Milliseconds } from '@cardano-sdk/core';
import { Connection, createConnectionObject } from '@cardano-ogmios/client';
import {
  MockOgmiosServerConfig,
  createMockOgmiosServer,
  listenPromise,
  mockGenesisConfig,
  serverClosePromise,
  waitForWsClientsDisconnect
} from '../mocks/mockOgmiosServer';
import { OgmiosObservableCardanoNode } from '../../src';
import { combineLatest, delay as delayEmission, firstValueFrom, mergeMap, of, take, toArray } from 'rxjs';
import { getRandomPort } from 'get-port-please';
import { logger } from '@cardano-sdk/util-dev';
import delay from 'delay';

const defaultConfig: MockOgmiosServerConfig = {
  chainSync: {
    requestNext: {
      responses: ['135.json']
    }
  },
  findIntersect: () => ({
    IntersectionFound: {
      point: 'origin',
      tip: 'origin'
    }
  }),
  maxConnections: 1,
  stateQuery: {
    eraSummaries: { response: { success: true } },
    genesisConfig: { response: { success: true } },
    systemStart: { response: { success: true } }
  }
};
const startMockOgmiosServer = async (port: number, configOverrides: Partial<MockOgmiosServerConfig> = {}) => {
  const mockServer = createMockOgmiosServer({
    ...defaultConfig,
    ...configOverrides
  });
  await listenPromise(mockServer, port);
  return mockServer;
};

describe('ObservableOgmiosCardanoNode', () => {
  let connection: Connection;

  beforeAll(async () => {
    connection = createConnectionObject({ port: await getRandomPort() });
  });

  describe('subscribing to all observables', () => {
    it('shares a single lazily created connection & emits well-formed data', async () => {
      const server = await startMockOgmiosServer(connection.port);
      const node = new OgmiosObservableCardanoNode({ connectionConfig$: of(connection) }, { logger });
      // Doesn't connect until any observable is subscribed
      await delay(100);
      expect(server.wss.clients.size).toBe(0);

      // Shares a single connection for all observables
      const res = await firstValueFrom(
        combineLatest([
          node.eraSummaries$,
          node.genesisParameters$,
          node.findIntersect(['origin']).pipe(mergeMap(({ chainSync$ }) => chainSync$))
        ])
      );
      // Data is well formed
      expect(res).toMatchSnapshot();
      // Websocket close() should be already called at this point,
      // but it takes some time to disconnect
      expect(server.wss.clients.size).toBe(1);

      // this will not resolve until all clients disconnect -
      // test will fail if OgmiosObservableCardanoNode
      // fails to cleanup connections when all observables are unsubscribed.
      await waitForWsClientsDisconnect(server, Milliseconds(500));
      await serverClosePromise(server);
    });
  });

  it('opaquely reconnects when connection is refused', async () => {
    const node = new OgmiosObservableCardanoNode({ connectionConfig$: of(connection) }, { logger });
    // Server is not started yet
    const result = firstValueFrom(node.genesisParameters$);
    // Observable emits once it can connect to the server
    await delay(100);
    const server = await startMockOgmiosServer(connection.port);
    // Once the server is started, it emits something that looks like genesis config
    expect(typeof (await result).epochLength).toBe('number');
    await serverClosePromise(server);
  });

  it('opaquely reconnects when websocket is closed by the server', async () => {
    const server = await startMockOgmiosServer(connection.port, {
      stateQuery: {
        ...defaultConfig.stateQuery,
        genesisConfig: {
          response: [
            {
              config: mockGenesisConfig,
              success: true
            },
            {
              config: {
                ...mockGenesisConfig,
                epochLength: mockGenesisConfig.epochLength + 1
              },
              success: true
            }
          ]
        }
      }
    });
    const node = new OgmiosObservableCardanoNode({ connectionConfig$: of(connection) }, { logger });
    const genesisParameters$ = node.genesisParameters$.pipe(take(2));
    const bothEmissions = firstValueFrom(genesisParameters$.pipe(toArray()));
    // connect and get genesis parameters
    await firstValueFrom(genesisParameters$);
    // simulate a non-"Normal Closure" event
    for (const client of server.wss.clients) {
      client.close(1001);
    }
    // it has to reconnect in order to emit the 2nd time
    const [{ epochLength: epochLength1 }, { epochLength: epochLength2 }] = await bothEmissions;
    expect(epochLength1).toEqual(mockGenesisConfig.epochLength);
    expect(epochLength2).toEqual(mockGenesisConfig.epochLength + 1);
    await serverClosePromise(server);
  });
  describe('healthCheck', () => {
    it('is ok if node is close to the network tip', async () => {
      const server = await startMockOgmiosServer(connection.port, {
        healthCheck: { response: { networkSynchronization: 0.999, success: true } }
      });
      const node = new OgmiosObservableCardanoNode({ connectionConfig$: of(connection) }, { logger });
      const res = await firstValueFrom(node.healthCheck$);
      expect(res.ok).toBe(true);
      await serverClosePromise(server);
    });
    it('simultaneous healthChecks share a single health request to ogmios', async () => {
      const server = await startMockOgmiosServer(connection.port, {
        healthCheck: { response: { networkSynchronization: 0.999, success: true } }
      });
      const node = new OgmiosObservableCardanoNode({ connectionConfig$: of(connection) }, { logger });
      const [res1, res2] = await firstValueFrom(combineLatest([node.healthCheck$, node.healthCheck$]));
      expect(res1.ok).toBe(true);
      expect(res1).toEqual(res2);
      expect(server.invocations.health).toBe(1);
      await serverClosePromise(server);
    });
    it('is not ok if node is not close to the network tip', async () => {
      const server = await startMockOgmiosServer(connection.port, {
        healthCheck: { response: { networkSynchronization: 0.8, success: true } }
      });
      const node = new OgmiosObservableCardanoNode({ connectionConfig$: of(connection) }, { logger });
      const res = await firstValueFrom(node.healthCheck$);
      expect(res.ok).toBe(false);
      await serverClosePromise(server);
    });
    it('is not ok when ogmios responds with an unknown result', async () => {
      const server = await startMockOgmiosServer(connection.port, {
        healthCheck: {
          response: { failWith: new CardanoNodeErrors.CardanoClientErrors.UnknownResultError(''), success: false }
        }
      });
      const node = new OgmiosObservableCardanoNode({ connectionConfig$: of(connection) }, { logger });
      const res = await firstValueFrom(node.healthCheck$);
      expect(res.ok).toBe(false);
      await serverClosePromise(server);
    });
    it('is not ok when connection is refused', async () => {
      const node = new OgmiosObservableCardanoNode({ connectionConfig$: of(connection) }, { logger });
      // Server is not started
      const result = await firstValueFrom(node.healthCheck$);
      expect(result.ok).toBe(false);
    });
    it('is ok when connectionConfig$ emits within "healthCheckTimeout" duration', async () => {
      const server = await startMockOgmiosServer(connection.port, {
        healthCheck: { response: { networkSynchronization: 0.999, success: true } }
      });
      const node = new OgmiosObservableCardanoNode(
        { connectionConfig$: of(connection).pipe(delayEmission(25)), healthCheckTimeout: Milliseconds(50) },
        { logger }
      );
      const result = await firstValueFrom(node.healthCheck$);
      expect(result.ok).toBe(true);
      await serverClosePromise(server);
    });
    it('is not ok when connectionConfig$ takes longer than "healthCheckTimeout" to emit', async () => {
      const server = await startMockOgmiosServer(connection.port, {
        healthCheck: { response: { networkSynchronization: 0.999, success: true } }
      });
      const node = new OgmiosObservableCardanoNode(
        { connectionConfig$: of(connection).pipe(delayEmission(50)), healthCheckTimeout: Milliseconds(25) },
        { logger }
      );
      const result = await firstValueFrom(node.healthCheck$);
      expect(result.ok).toBe(false);
      await serverClosePromise(server);
    });
  });
});
