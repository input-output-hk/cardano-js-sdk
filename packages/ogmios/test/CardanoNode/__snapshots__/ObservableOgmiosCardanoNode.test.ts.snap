// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObservableOgmiosCardanoNode subscribing to all observables shares a single lazily created connection & emits well-formed data 1`] = `
Array [
  Array [
    Object {
      "parameters": Object {
        "epochLength": 21600,
        "slotLength": 20000,
      },
      "start": Object {
        "slot": 0,
        "time": 2017-09-23T21:44:51.000Z,
      },
    },
    Object {
      "parameters": Object {
        "epochLength": 432000,
        "slotLength": 1000,
      },
      "start": Object {
        "slot": 1598400,
        "time": 2018-09-28T21:44:51.000Z,
      },
    },
  ],
  Object {
    "activeSlotsCoefficient": 0.05,
    "epochLength": 86400,
    "maxKesEvolutions": 120,
    "maxLovelaceSupply": 45000000000000000n,
    "network": "testnet",
    "networkId": 0,
    "networkMagic": 2,
    "securityParameter": 432,
    "slotLength": 1,
    "slotsPerKesPeriod": 86400,
    "systemStart": 2022-08-09T00:00:00.000Z,
    "updateQuorum": 5,
  },
  Object {
    "block": Object {
      "body": Array [
        Object {
          "auxiliaryData": undefined,
          "body": Object {
            "auxiliaryDataHash": undefined,
            "certificates": Array [],
            "collaterals": Array [
              Object {
                "index": 0,
                "txId": "ae85d245a3d00bfde01f59f3c4fe0b4bfae1cb37e9cf91929eadcea4985711de",
              },
              Object {
                "index": 0,
                "txId": "ee155ace9c40292074cb6aff8c9ccdd273c81648ff1149ef36bcea6ebb8a3e25",
              },
            ],
            "fee": 648n,
            "inputs": Array [],
            "mint": Map {},
            "outputs": Array [],
            "requiredExtraSignatures": Array [],
            "scriptIntegrityHash": "bb30a42c1e62f0afda5f0a4e8a562f7a13a24cea00ee81917b86b89e801314aa",
            "validityInterval": Object {
              "invalidBefore": 26215,
              "invalidHereafter": undefined,
            },
            "withdrawals": Array [],
          },
          "id": "a278b540fb192eb8276a0d49004968076c9f9f026fa8727c3f9981a4aac25cc9",
          "inputSource": "inputs",
          "witness": Object {
            "bootstrap": Array [],
            "datums": Array [
              "40",
            ],
            "redeemers": Array [
              Object {
                "data": "434d357a",
                "executionUnits": Object {
                  "memory": 8959327235984229000n,
                  "steps": 4473903472119932000n,
                },
                "index": 2,
                "purpose": "certificate",
              },
              Object {
                "data": "9fd866821901739fd866821902dd9f2341e14478c0fcf5ff22d866821903e39f438002ab0104ffa1448858788900ff029fa040ffff",
                "executionUnits": Object {
                  "memory": 4858311980956467000n,
                  "steps": 1434644070377693400n,
                },
                "index": 1,
                "purpose": "withdrawal",
              },
            ],
            "scripts": Array [
              Object {
                "__type": "plutus",
                "bytes": "450100002601",
                "version": 1,
              },
            ],
            "signatures": Map {},
          },
        },
      ],
      "fees": 648n,
      "header": Object {
        "blockNo": 2,
        "hash": "bb30a42c1e62f0afda5f0a4e8a562f7a13a24cea00ee81917b86b89e801314aa",
        "slot": 8,
      },
      "issuerVk": "5effdd0f8a1c75abfbd837ced4bc50fe8e42aded5d5a310addcba6eccb495d78",
      "previousBlock": "03170a2e7597b7b7e3d84c05391d139a62b157e78786d8c082f29dcf4c111314",
      "size": 757,
      "totalOutput": 0n,
      "txCount": 1,
      "vrf": "vrf_vk15rekjgqyvguv2eggspcmdda7xkgrf2tdmj2kgv22ejcwwujhzc7sfl7zpz",
    },
    "eventType": 0,
    "requestNext": [Function],
    "tip": Object {
      "blockNo": 9501533,
      "hash": "c190ad992ae264008253235c98e3ccdcd78d5d95f74f317f9c8af0482cb04b0b",
      "slot": 11119,
    },
  },
]
`;
